DYN-6569 Extended Characters Md2HTML Fix

RobertGlobant20 opened the pull request:
### Purpose

Fixing problem when displaying extended characters in DocumentationBrowser
When sending extended characters to the Md2HTML tool (using stdin and stdout) was replacing the characters by other weird symbols then for fixing this problem I'm encoding the info to base64 when sending and decoding the info when getting the html (from Md2HTML).

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [X] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Fixing problem when displaying extended characters in DocumentationBrowser

### Reviewers

@QilongTang 

### FYIs

@mjkkirschner

RobertGlobant20 commented:
SweepAsSurface - Czech
![tR1Rz728pm](https://github.com/DynamoDS/Dynamo/assets/61755417/d5660555-9696-42fb-bde7-3089e171faf6)

RobertGlobant20 commented:
ImportExcel- Czech
![FtiYHc9C0Y](https://github.com/DynamoDS/Dynamo/assets/61755417/d709a09a-76b6-4468-8f86-50fcb2306cc7)

RobertGlobant20 commented:
SweepAsSurface - 한국인
![RyPw2gelC9](https://github.com/DynamoDS/Dynamo/assets/61755417/85163ae2-95bd-400c-a754-25841581c249)

RobertGlobant20 commented:
ImportExcel - 한국인
![wj2rkuJa5D](https://github.com/DynamoDS/Dynamo/assets/61755417/bda0f45f-4c01-4ba6-9d62-d395494500a8)

QilongTang commented on src/DynamoUtilities/CLIWrapper.cs:
```diff
@@ -162,6 +165,47 @@ protected virtual string GetData(int timeoutms, Func<string> mockReadLine = null
+                            var decodedString = Encoding.UTF8.GetString(base64Line);
+                            htmlContent = htmlContent.Replace(encodedString, decodedString);
+                        }
+                        catch (Exception ex)
```
You can do `catch (Exception)` if you dont plan to print out ex info

RobertGlobant20 replied:
Updated the catch in the next commit: https://github.com/DynamoDS/Dynamo/pull/15377/commits/33430d9cb836caa94d02429978d28021579d2319

QilongTang commented on src/DynamoUtilities/CLIWrapper.cs:
```diff
@@ -162,6 +165,47 @@ protected virtual string GetData(int timeoutms, Func<string> mockReadLine = null
             return completedTask == readStdOutTask ? readStdOutTask.Result : string.Empty;
         }
 
+        private void DecodeHTMLContent(ref string htmlContent)
```
These two functions maybe better belong to `Md2Html` class?

RobertGlobant20 replied:
The two methods were moved to the Md2Html class but I had to do the methods static (because otherwise are not visible in CLIWrapper.cs)
https://github.com/DynamoDS/Dynamo/pull/15377/commits/33430d9cb836caa94d02429978d28021579d2319

QilongTang commented on src/DynamoUtilities/Md2Html.cs:
```diff
@@ -76,6 +79,37 @@ internal string ParseMd2Html(string mdString, string mdPath)
+            EncodeBase64(ref mdContent, @"#+\s(.*?\n)");
+        }
+
+        private void EncodeBase64(ref string mdContent, string regEx)
```
Would you put some comments to the function? And do you think we can add a test case?

RobertGlobant20 replied:
Added comments for the method and the parameters received.
Related to the test case, I think there is no way to check if the content of the DocumentationBrowser is showing the expected characters in a specific language, probably something that we can do is trying to process some md files in different languages and then check that the generated html has a similar content that the md file.
What do you think?
https://github.com/DynamoDS/Dynamo/pull/15377/commits/33430d9cb836caa94d02429978d28021579d2319

QilongTang replied:
@RobertGlobant20 yup, totally fine. No need to switch Dynamo to a different locale, loading md file in a different locale and check content is a fine test case

RobertGlobant20 replied:
I've added a unit test that validates that the extended characters (in en-US, ko-KR, cs-CZ) are processed correctly by the Md2Html.exe tool.
https://github.com/DynamoDS/Dynamo/pull/15377/commits/1d7cd8d0afae6f3efa75f9e50509c28ae88ad8e5

QilongTang commented on src/DynamoUtilities/Md2Html.cs:
```diff
@@ -76,6 +79,81 @@ internal string ParseMd2Html(string mdString, string mdPath)
             return GetData(processCommunicationTimeoutms);
         }
 
+        private void EncodeMDContent(ref string mdContent)
```
How about making these two functions here also static?

RobertGlobant20 replied:
This functions were also changed to be static.
commit: https://github.com/DynamoDS/Dynamo/pull/15377/commits/1d7cd8d0afae6f3efa75f9e50509c28ae88ad8e5

QilongTang reviewed (commented):
LGTM pending a comment and a potential test case

QilongTang commented:
Regression:
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionContentTests.CheckThatExtendedCharactersAreCorrectlyInHTML("cs-CZ")](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16275/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionContentTests/CheckThatExtendedCharactersAreCorrectlyInHTML__cs_CZ__/)
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionContentTests.CheckThatExtendedCharactersAreCorrectlyInHTML("ko-KR")](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16275/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionContentTests/CheckThatExtendedCharactersAreCorrectlyInHTML__ko_KR__/)

QilongTang commented:
@RobertGlobant20 Regressions:
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionTests.CanCreateNodeDocumenationHtmlFromNodeAnnotationEventArgsWithPackageNodeWithAddtionalDocumentation](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16278/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionTests/CanCreateNodeDocumenationHtmlFromNodeAnnotationEventArgsWithPackageNodeWithAddtionalDocumentation/)
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionTests.CanCreatePackageNodeDocumentationAndLoadImages](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16278/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionTests/CanCreatePackageNodeDocumentationAndLoadImages/)

RobertGlobant20 commented:
> @RobertGlobant20 Regressions: [DynamoCoreWpfTests.DocumentationBrowserViewExtensionTests.CanCreateNodeDocumenationHtmlFromNodeAnnotationEventArgsWithPackageNodeWithAddtionalDocumentation](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16278/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionTests/CanCreateNodeDocumenationHtmlFromNodeAnnotationEventArgsWithPackageNodeWithAddtionalDocumentation/) [DynamoCoreWpfTests.DocumentationBrowserViewExtensionTests.CanCreatePackageNodeDocumentationAndLoadImages](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16278/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionTests/CanCreatePackageNodeDocumentationAndLoadImages/)

I've removed the code for encoding and decoding so this tests should be passing and also the test that I added.

QilongTang commented:
hi @RobertGlobant20 Still failing it seems:
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionContentTests.CheckThatExtendedCharactersAreCorrectlyInHTML("en-US")](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16281/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionContentTests/CheckThatExtendedCharactersAreCorrectlyInHTML__en_US__/)
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionContentTests.CheckThatExtendedCharactersAreCorrectlyInHTML("cs-CZ")](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16281/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionContentTests/CheckThatExtendedCharactersAreCorrectlyInHTML__cs_CZ__/)
[DynamoCoreWpfTests.DocumentationBrowserViewExtensionContentTests.CheckThatExtendedCharactersAreCorrectlyInHTML("ko-KR")](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16281/testReport/junit/DynamoCoreWpfTests/DocumentationBrowserViewExtensionContentTests/CheckThatExtendedCharactersAreCorrectlyInHTML__ko_KR__/)
